pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -255,6 +281,88 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
 	return pod
 }
 
+// newVersionedStatefulSetPod creates a new Pod for a StatefulSet. currentSet is the representation of the set at the
+// current revision. updateSet is the representation of the set at the updateRevision. currentRevision is the name of
+// the current revision. updateRevision is the name of the update revision. ordinal is the ordinal of the Pod. If the
+// returned error is nil, the returned Pod is valid.
+func newVersionedStatefulSetPod(currentSet *apps.StatefulSet, updateSet *apps.StatefulSet, currentRevision string, updateRevision string, ordinal int) *v1.Pod {
+	if (currentSet.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		ordinal < int(currentSet.Status.CurrentReplicas)) ||
+		(currentSet.Spec.UpdateStrategy.Type == apps.PartitionStatefulSetStrategyType &&
+			ordinal < int(currentSet.Spec.UpdateStrategy.Partition.Ordinal)) {
+		pod := newStatefulSetPod(currentSet, ordinal)
+		setPodRevision(pod, currentRevision)
+		return pod
+	}
+	pod := newStatefulSetPod(updateSet, ordinal)
+	setPodRevision(pod, updateRevision)
+	return pod
+}
+
+// getPatch returns a strategic merge patch that can be applied to restore a StatefulSet to a
+// previous version. If the returned error is nil the patch is valid.
+func getPatch(set *apps.StatefulSet) ([]byte, error) {
+	str, err := runtime.Encode(patchCodec, set)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	json.Unmarshal([]byte(str), &raw)
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
+	template := spec["template"].(map[string]interface{})
+	specCopy["template"] = template
+	template["$patch"] = "replace"
+	objCopy["spec"] = specCopy
+	patch, err := json.Marshal(objCopy)
+	return patch, err
+}
+
+// newRevision creates a new ControllerRevision containing a patch that can convert a StatefulSet to an equivalent state to set.
+// The Revision of the returned ControllerRevision is set to Revision. If the returned error is nil. the returned
+// ControllerRevision is valid.
+func newRevision(set *apps.StatefulSet, revision int64) (*apps.ControllerRevision, error) {
+	patch, err := getPatch(set)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return history.NewControllerRevision(set,
+		controllerKind,
+		selector,
+		runtime.RawExtension{Raw: patch},
```

Comment:
this is confusing and at the same time interesting to me as a concept :-) I have not seen this pattern before, but it seems like what you are doing is keeping a strategic merge patch in each revision history , so that any time we want to restore the StatefulSet to that revision, you just use the ready to apply patch from the Raw field to restore it. I think this is worth some more comment